test/nodetool/test_toppartitions.py

```diff
@@ -0,0 +1,140 @@
+
+
+@pytest.mark.parametrize("empty_samplings", [True, False])
+def test_toppartitions(nodetool, request, empty_samplings):
+
+    if empty_samplings:
+        samplings = {
+            'read_cardinality': 0,
+            'write_cardinality': 0,
+        }
+        response = samplings
+    else:
+        samplings = {
+            'read_cardinality': 5,
+            'read': [Record('Russell Westbrook', 100, 3),
+                     Record('Jermi Grant',       25,  1),
+                     Record('Victor Oladipo',    17,  0),
+                     Record('Andre Roberson',    1,   0),
+                     Record('Steven Adams',      1,   0)],
+            'write_cardinality': 5,
+            'write': [Record('Russell Westbrook', 101, 3),
+                      Record('Jermi Grant',       26,  1),
+                      Record('Victor Oladipo',    18,  0),
+                      Record('Andre Roberson',    2,   0),
+                      Record('Steven Adams',      2,   0)],
+        }
+        response = {
+            'read_cardinality': samplings['read_cardinality'],
+            'read': [record.to_json() for record in samplings['read']],
+            'write_cardinality': samplings['write_cardinality'],
+            'write': [record.to_json() for record in samplings['write']],
+        }
+    duration = 50000
+    capacity = 512
+    list_size = 20
+    table_filters = 'ks:cf'
+    options = {
+        "--duration": duration,
+        "-s": capacity,
+        "-k": list_size,
+        "--cf-filters": table_filters,
+    }
+    args = []
+    for name, value in options.items():
+        args.extend([name, str(value)])
+    params = {
+        'duration': str(duration),
+        'table_filters': 'ks:cf',
+        'capacity': str(capacity),
+    }
+    if request.config.getoption("nodetool") == "scylla":
+        # scylla sends list_size, while cassandra's nodetool does not.
+        params['list_size'] = str(list_size)
+    actual_output = nodetool("toppartitions", *args, expected_requests=[
+        expected_request("GET", "/storage_service/toppartitions",
+                         params=params,
+                         response=response),
+    ])
+
+    expected_output = ''
+    first = True
+    for operation in ('read', 'write'):
+        operation_upper = operation.upper()
+        cardinality = samplings[f'{operation}_cardinality']
+        expected_output += f'''\
+{operation_upper}S Sampler:
+  Cardinality: ~{cardinality} ({capacity} capacity)
+  Top {list_size} partitions:
+'''
+        if operation not in samplings:
+            expected_output += "\tNothing recorded during sampling period...\n"
+            continue
+
+        expected_output += Record.header()
+        for record in samplings[operation]:
+            expected_output += record.format()
+        if first:
+            expected_output += '\n'
+            first = False
+
+    assert normalize_samplings(actual_output) == normalize_samplings(expected_output)
```

Comment:
I see you do some sanity checking of args, please add negative tests, where you pass the wrong params and scylla-nodetool exits with the expected error message. We only need such tests for validation the client (nodetool) is expected to do.

Reply:
> We only need such tests for validation the client (nodetool) is expected to do.

could you please elaborate a little bit on this? do you suggest that we add these negative tests for C*'s nodetool? if so, what's the point?

Reply:
No, I suggest we add these for both implementations, so we have a test case for all the different command-line validation related errors they emit.
If some validation is only done by scylla-nodetool, we should only add a test for it (use `scylla_only` fixture).

Reply:
I meant we don't need to add tests for validations done by the server itself (ScyllaDB Rest API).

Reply:
i see. on the contrary, i don't intend to do more than what C*'s client does unless i believe that the optimisation/improvement is simple and straightforward. the sanity checks mirrors its behavior at the client side. and i am not adding the tests on behalf of server. or i am not aware of that server does the same check. if you still think so, could you point it out?

Reply:
I agree, this is what I do too. I add checks if C* has them, but nothing more, and cover these with tests which trigger the checks.
We don't need to worry about server-side checks here because we use a mock server. We have end-to-end tests for those in dtests.git (I think).